### PR TITLE
os/bluestore: Copy objects to new zone in zoned_clean_zone to truncate zone

### PR DESCRIPTION
**Pull request by Rishabh4275**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by Rishabh4275** [comment]
@agayev 
@ifed01 can you please review this?

**Comment by agayev** [comment]
@ifed01 can you please see if the proposed list of steps makes sense?

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by Rishabh4275** [comment]
jenkins test Docs: build check

**Comment by agayev** [comment]
@ifed01 would appreciate if you could take a look. Thanks.

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -12399,6 +12399,40 @@ void BlueStore::_zoned_clean_zone(uint64_t zone_num) {
...
+  {
+    std::string k = it->key();
+    bufferlist bl = it->value();
+    auto p = bl.cbegin();
+    int64_t offset;
+    ::decode(offset, p);
+    dout(40) << __func__ << " Copying object with key (zone_num + oid): " << k << " and offset: " << offset << dendl;
+    ghobject_t oid;
+    int r = get_key_object(k, &oid);
+    CollectionRef c = _get_collection(coll_t::meta());
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:12417]: one should use collection related to the object not 'meta' one here. Apparently you should persist that somehow - the issue is that collection might split/merge over time...

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -12399,6 +12399,40 @@ void BlueStore::_zoned_clean_zone(uint64_t zone_num) {
...
+    OnodeRef o = c->get_onode(oid, false);
+    o->extent_map.fault_range(db, 0, o->onode.size);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:12419]: this is redundant, _do_read() calls it.

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -12399,6 +12399,40 @@ void BlueStore::_zoned_clean_zone(uint64_t zone_num) {
   dout(10) << __func__ << " cleaning zone " << zone_num << dendl;
   // TODO: (1) copy live objects from zone_num to a new zone, (2) issue a RESET
   // ZONE operation to the device for the corresponding zone.
+
+  zone_state_t zone_state;
+  std::string pfx = _zoned_get_prefix(zone_num * bdev->get_zone_size());
+  KeyValueDB::Iterator it = db->get_iterator(pfx, KeyValueDB::ITERATOR_NOCACHE);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:12405]: IMO you should revise naming scheme a bit for the same of uniformity with other stuff.
Prefix to be a single char field, e.g. 'G'
And the key itself to includes zone + oid + whatever...
Hence one will need to call it->lower_bound(key(zone)) after get_iterator to locate the first record belonging to a desired zone.

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -12399,6 +12399,40 @@ void BlueStore::_zoned_clean_zone(uint64_t zone_num) {
...
+    ::decode(offset, p);
+    dout(40) << __func__ << " Copying object with key (zone_num + oid): " << k << " and offset: " << offset << dendl;
+    ghobject_t oid;
+    int r = get_key_object(k, &oid);
+    CollectionRef c = _get_collection(coll_t::meta());
+    OnodeRef o = c->get_onode(oid, false);
+    o->extent_map.fault_range(db, 0, o->onode.size);
+    ceph_assert(offset == o->zoned_get_ondisk_starting_offset());
+    //Should I use _read here and the flag = CEPH_OSD_OP_FLAG_FADVISE_NOCACHE
+    r = _do_read(c.get(), o, 0, o->onode.size, bl, 0);
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:12422]: There are some concerns about the code starting from this line:
1) Object isn't locked at this point - hence possible races with other users.
2) reading and writing are not within an atomic transaction.
3) Reading in a single shot is doubtful (from RAM usage point of view) when dealing with huge objects. Suggest to perform using limited block sizes. Preferably by using underlying continuous physical extents  which are of limited size by their nature and benefit from the lack of fragmentation. Maybe multiple pextents to be processed in a single shot (atomic transaction) shot.
4) One doesn't need to make any object 'clones'. This should be just an atomic move operation on object's data (or a set of atomic operations). I'd dream about it's implementation as a standalone method enumerating onode's extents (or blobs?) and hence effectively determining blocks to process.
5) Please remind me whether we pin object exclusively to a specific zone? Can it have pextents from multiple zones? If so - we might need to handle that here...

**Comment by Rishabh4275** [comment]
jenkins test make check

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (agayev) [reply src/os/bluestore/BlueStore.cc:12431]: @ifed01 Hmmm, so we should persist collection id as well. I'm being lazy here, but with the rados tool, I can write an object and then retrieve the object using just its name. So what am I missing here? Why can't we do the same using the low-level API?

File: src/os/bluestore/BlueStore.cc (hunk 3 above)
Comment (agayev) [reply src/os/bluestore/BlueStore.cc:12419]: This one is resolved in https://github.com/ceph/ceph/pull/41845

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (ifed01) [reply src/os/bluestore/BlueStore.cc:12431]: Onodes are addressable via a <cid, oid> pair, single oid isn't enough. And every onode is bound to a specific collection hence one needs to get proper collection reference via _get_collection(cid) call to access that onode first. 

The above _get_collection(coll_t::meta()) returns just a "meta" collection which keeps just a few particular onodes, not all of them. 
So one needs to preserve onode's cid when persisting onode information in _zoned_update_cleaning_metadata() and reuse that cid here...

**Pull request closed without merging**